# Changelog

## [0.2.2](https://github.com/clinical-genomics-uppsala/poppy_uppsala/compare/v0.2.1...v0.2.2) (2025-06-03)


### Bug Fixes

* merge AF after VT normalization for complex variants found by pindel ([679b39b](https://github.com/clinical-genomics-uppsala/poppy_uppsala/commit/679b39b20c019162f59d20ff508dd0a5ef2be9a8))
* snakefmt linting ([bc14914](https://github.com/clinical-genomics-uppsala/poppy_uppsala/commit/bc149148eeba5263c783032ccd313457b04e0be8))

## [0.2.1](https://github.com/clinical-genomics-uppsala/poppy_uppsala/compare/v0.2.0...v0.2.1) (2025-06-03)


### Bug Fixes

* hotspot genes in excel report ([e57ce10](https://github.com/clinical-genomics-uppsala/poppy_uppsala/commit/e57ce107fe4a0c8e816a8eb871a0a962046e5691))
* removed pureCN ([3bc3a16](https://github.com/clinical-genomics-uppsala/poppy_uppsala/commit/3bc3a166265c5f9ba96462c7bfb22bcdc7f4903c))

## [0.2.0](https://github.com/clinical-genomics-uppsala/poppy_uppsala/compare/v0.1.2...v0.2.0) (2025-04-25)


### Features

* container for venv with bamsnap ([cb824d5](https://github.com/clinical-genomics-uppsala/poppy_uppsala/commit/cb824d5ae6fb1e657b1790b554760cfaaf1c70be))


### Bug Fixes

* linting container directive ([351504c](https://github.com/clinical-genomics-uppsala/poppy_uppsala/commit/351504c0761bc746ac236fb0eeb6bf29f82d38c4))
* linting snakefmt ([3f33f43](https://github.com/clinical-genomics-uppsala/poppy_uppsala/commit/3f33f43052a50452a68325ea133e288e52dafe41))
* linting snakefmt ([eb913af](https://github.com/clinical-genomics-uppsala/poppy_uppsala/commit/eb913afb7e49ff17e55631953cc490106e175b18))
* linting snakefmt ([beca1f4](https://github.com/clinical-genomics-uppsala/poppy_uppsala/commit/beca1f43909b03096588724893b58f79072cbbbd))
* merge AF after normalization if 2 variants at same pos ([e328e9e](https://github.com/clinical-genomics-uppsala/poppy_uppsala/commit/e328e9e51b70d95bd29c3c146b3c8ed65fcda5e2))

## 0.1.0 (2025-01-31)


### Features

* add artifacts ([41be3f8](https://github.com/clinical-genomics-uppsala/poppy_uppsala/commit/41be3f8ce8046285900fd6a7476240fb7828e9f2))
* add xlsx snv, indels, known sheets ([c560975](https://github.com/clinical-genomics-uppsala/poppy_uppsala/commit/c560975955ca3f3a620647bb7c6787dd61b869ab))
* merge 'develop' ([2d0e37e](https://github.com/clinical-genomics-uppsala/poppy_uppsala/commit/2d0e37e5a48d5d0a7bb0a93ee89703f647788ba5))
* remove shortlist and replace with panels using bcftools include ([c6a010d](https://github.com/clinical-genomics-uppsala/poppy_uppsala/commit/c6a010d4ac056ce9493667e6179ad1ef356f9e63))
* start xlsxfile ([631a733](https://github.com/clinical-genomics-uppsala/poppy_uppsala/commit/631a733100e451d9b7372e4146a06c603418cb3f))


### Bug Fixes

* add missig panels[panel] ([5cdcc24](https://github.com/clinical-genomics-uppsala/poppy_uppsala/commit/5cdcc247826cbaaf2e79432e134d7db13b729036))
* add missing lib, and change varible chr to chrom ([e0d4cc9](https://github.com/clinical-genomics-uppsala/poppy_uppsala/commit/e0d4cc926c5719eac0dd30437cab74a42be6a493))
* change from intervals to interval_list ([ee9d017](https://github.com/clinical-genomics-uppsala/poppy_uppsala/commit/ee9d017d19320b885e9f46f9ea29f17a0ef1e7c5))
* correct config file  for snakemake-dry-run.yaml ([453a83c](https://github.com/clinical-genomics-uppsala/poppy_uppsala/commit/453a83c01835458c351359282d0bbd8f0cfc02e9))
* lock numpy to v1.26.4 ([ef04cc8](https://github.com/clinical-genomics-uppsala/poppy_uppsala/commit/ef04cc8f8e3d4da026f0d7e5b78a3143818a7973))
* make sure copy results can handle folders ([449fd98](https://github.com/clinical-genomics-uppsala/poppy_uppsala/commit/449fd983087ba3da8a960aae63e82f59f3b57a20))
* update common to match poppys ([dcb6cc6](https://github.com/clinical-genomics-uppsala/poppy_uppsala/commit/dcb6cc654e42c4fb6286f69b66634685ec9b0554))
* update linting test ([b8f8d55](https://github.com/clinical-genomics-uppsala/poppy_uppsala/commit/b8f8d555db9f968b12f1f65fa70c25c340955760))
* update requirements.test.txt ([aac38c9](https://github.com/clinical-genomics-uppsala/poppy_uppsala/commit/aac38c961dc9a3e0ae858153b8853186271534df))
* update requirements.txt ([192099a](https://github.com/clinical-genomics-uppsala/poppy_uppsala/commit/192099a598e3a28b4c3b668e0204ddafdca4954e))
* update smart_open version to lower than 7.0.0 ([8ece5c9](https://github.com/clinical-genomics-uppsala/poppy_uppsala/commit/8ece5c9601fa31316059cd73bf17291d5a3c26c5))
* update snakefmt.yaml ([959033b](https://github.com/clinical-genomics-uppsala/poppy_uppsala/commit/959033be8a356f3d8629b16d522a7c2da51dfbf5))
* update so not all col set to str when loading samples.tsv ([36207e3](https://github.com/clinical-genomics-uppsala/poppy_uppsala/commit/36207e3a45ad9fe671e5c5fb683032f5d35d573a))


### Documentation

* add default addded hydra docs ([0976c04](https://github.com/clinical-genomics-uppsala/poppy_uppsala/commit/0976c04ab3667d1d28beb393943d9850eb4f747e))
* schema updates ([3580b41](https://github.com/clinical-genomics-uppsala/poppy_uppsala/commit/3580b419d633a8353aa23b2899c2d164f7c6b437))
